rust/kernel/cpu.rs

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! CPU utilities for SMP systems.
+//!
+//! C header: [`include/linux/smp.h`](../../../../include/linux/smp.h).
+//!
+//! Note: This module is a work in progress.
+
+use crate::{bindings, error::code::*, prelude::*};
+use core::marker::PhantomData;
+
+/// Represents an SMP CPU by its zero-based kernel processor id.
+///
+/// Cpu values can be safely passed between threads (and across physical CPUs) as they are
+/// merely an integer index representing a CPU. All actual interaction with CPUs is late-
+/// bound.
+#[derive(Copy, Clone)]
+pub struct Cpu(i32);
+
+struct FnCall<F, R>
+where
+    F: FnOnce() -> R,
+{
+    out: Option<R>,
+    func: Option<F>,
+}
+
+unsafe extern "C" fn call_fn<F, R>(fn_box: *mut core::ffi::c_void)
+where
+    F: FnOnce() -> R,
+{
+    // SAFETY: The casting here is the invesrse of the cast to *mut c_void that occurs in Cpu::call
+    // below. That call waits for this call to complete ensuring fn_call is live for the duration.
+    let fn_call = unsafe { &mut *(fn_box as *mut FnCall<F, R>) };
+    let func = fn_call.func.take().unwrap();
+    fn_call.out = Some(func());
+}
+
+impl Cpu {
+    /// Constructs a new instance referring to a specific processor by id.
+    ///
+    /// This function checks that the passed id is a valid _potential_ processor id, but does
+    /// not check that the processor is actually present or online. Manipulation of absent
+    /// or offline processors will fail with reasonable errors returned.
+    pub fn from_id(id: i32) -> Result<Cpu> {
+        if id >= Self::num_possible() {
+            return Err(E2BIG);
+        } else if id < 0 {
+            return Err(EINVAL);
+        }
+        Ok(Cpu(id))
+    }
+
+    /// Returns the number of processors that are possible (including onlining and hotplug).
+    pub fn num_possible() -> i32 {
+        // SAFETY: The number of possible CPUs is fixed early at boot and does not change in
+        // response to processors coming online, offline, or hotplug events.
+        unsafe { bindings::nr_cpu_ids as i32 }
+    }
+
+    /// Lock the running thread to the current CPU and prevent preemption.
+    pub fn lock_current() -> Guard {
+        Guard::new()
+    }
+
+    /// Get the processor id of this CPU.
+    pub fn id(&self) -> i32 {
+        self.0
+    }
+
+    /// Call a function on the CPU identified by `self`.
+    pub fn call<R, F>(&self, func: F) -> Result<R>
+    where
+        F: FnOnce() -> R + Send,
+        R: Send,
+    {
+        let mut fn_call = FnCall {
+            out: None,
+            func: Some(func),
+        };
+        // SAFETY: smp_call_function_single is invoked with wait == 1 (true) ensuring that fn_call
+        // outlives this FFI invocation. func's returned value is moved into fn_call.out and then
+        // returned here. The Send type constraint ensures func and its returned value can safely
+        // be moved between threads/CPUs.
+        unsafe {
```

Comment:
nit: we try to make `unsafe` blocks as small as possible. In this case, it should only be around the `smp_call_function_single` call.